sonar-batch-protocol/src/main/java/org/sonar/batch/protocol/ProtobufUtil.java

```diff
@@ -44,6 +39,43 @@ private ProtobufUtil() {
     }
   }
 
+  public static <MESSAGE extends Message> Iterable<MESSAGE> readFileMessages(File file, final Parser<MESSAGE> parser) {
+    try {
+      final InputStream inputStream = new BufferedInputStream(new FileInputStream(file));
+      return new Iterable<MESSAGE>() {
+        @Override
+        public Iterator<MESSAGE> iterator() {
+          return new Iterator<MESSAGE>() {
+
+            private MESSAGE currentMessage;
+
+            @Override
+            public boolean hasNext() {
+              try {
+                currentMessage = parser.parseDelimitedFrom(inputStream);
+              } catch (InvalidProtocolBufferException e) {
+                throw new IllegalStateException("Failed to read input stream", e);
+              }
+              return currentMessage != null;
+            }
+
+            @Override
+            public MESSAGE next() {
+              return currentMessage;
```

Comment:
With this design it is mandatory to call hasNext() before each call to next(). And it is not possible to call hasNext() several times. Later someone can be surprised by this behavior that is not the contract of Iterator.
iterator.hasNext();
iterator.next() -> return first element
iterator.hasNext();
iterator.hasNext();
iterator.next() -> return third element